e2e/playwright/conftest.py

```diff
@@ -0,0 +1,365 @@
+            self._stdout_file = None
+
+
+def is_port_available(port: int, host: str) -> bool:
+    """Check if a port is available on the given host."""
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+        return sock.connect_ex((host, port)) != 0
+
+
+def find_available_port(
+    min_port: int = 10000,
+    max_port: int = 65535,
+    max_tries: int = 50,
+    host: str = "localhost",
+) -> int:
+    """Find an available port on the given host."""
+    for _ in range(max_tries):
+        selected_port = randint(min_port, max_port)
+        if is_port_available(selected_port, host):
+            return selected_port
+    raise RuntimeError("Unable to find an available port.")
+
+
+def is_app_server_running(port: int, host: str = "localhost") -> bool:
+    """Check if the app server is running."""
+    try:
+        return requests.get(f"http://{host}:{port}/", timeout=1).status_code == 200
+    except Exception:
+        return False
+
+
+def wait_for_app_server_to_start(port: int, timeout: int = 5) -> bool:
+    """Wait for the app server to start."""
+    print(f"Waiting for app to start... {port}")
+    start_time = time.time()
+    while not is_app_server_running(port):
+        time.sleep(3)
+        if time.time() - start_time > 60 * timeout:
+            return False
+    return True
+
+
+def wait_for_app_loaded(page: Page):
+    """Wait for the app to fully load."""
+    # Wait until we know the script has started.
+    page.wait_for_selector(
+        "[data-testid='stAppViewContainer']", timeout=20000, state="attached"
+    )
+
+    page.wait_for_selector(
+        "[data-testid='block-container']", timeout=20000, state="attached"
+    )
+
+    # Give the app a little more time to render everything
+    time.sleep(0.25)
+    # Check that "Please wait..." does not exist within the 'stAppViewContainer'
+    app_view_container = page.query_selector("[data-testid='stAppViewContainer']")
+    if (
+        app_view_container
+        and app_view_container.inner_text().find("Please wait...") != -1
+    ):
+        pytest.fail("Found 'Please wait...' inside the stAppViewContainer")
+    # Wait until the script is no longer running.
+    page.wait_for_selector(
+        "[data-testid='stStatusWidget']", timeout=20000, state="detached"
+    )
+    # Give the app a little more time to render everything
+    time.sleep(0.5)
+
+
+@pytest.fixture(scope="module")
+def app_port() -> int:
+    """Fixture that returns an available port on localhost."""
+    return find_available_port()
+
+
+@pytest.fixture(scope="module", autouse=True)
+def app_server(
+    app_port: int, request: FixtureRequest
+) -> Generator[AsyncSubprocess, None, None]:
+    """Fixture that starts and stops the Streamlit app server."""
+    module_name = request.module.__name__
+    streamlit_proc = AsyncSubprocess(
+        [
+            "streamlit",
+            "run",
+            module_name.replace("_test", ".py"),  # TODO: a better way?
```

Comment:
I would consider whether it is worth configuring the assertion at the module level using the variable `pytestmark`.
```
pytestmark = pytest.mark.usefixtures("cleandir")
```

Reply:
I'm not 100% sure what you mean by this comment. Is it referring to the specific line you marked above?

Reply:
In the case of testing custom components, I would not like the script file path to be dynamically determined, but it should be determined by the test author, so as not to force a specific test/directory convention.


Something similar to the code below, but I'll deal with this problem later.
```
pytestmark = pytest.mark.e2e_test("st_arrow_dataframe_column_types.py")
```
This is not needed in this PR.